image_resources.cpp

```diff
@@ -130,5 +132,225 @@ bool ImageResource::resIDHasDescriptor(uint16_t resID)
       return false;
   }
 }
+bool is_valid_class_type(const uint32_t class_id) {
+    // we don't want to cast the class_id to OSTypeKey to avoid UB
+    switch (class_id) {
+    case (uint32_t)OSTypeKey::Alias:
+    case (uint32_t)OSTypeKey::Boolean:
+    case (uint32_t)OSTypeKey::ClassType:
+    case (uint32_t)OSTypeKey::Descriptor:
+    case (uint32_t)OSTypeKey::Double:
+    case (uint32_t)OSTypeKey::Enumerated:
+    case (uint32_t)OSTypeKey::GlobalClass:
+    case (uint32_t)OSTypeKey::GlobalObject:
+    case (uint32_t)OSTypeKey::LargeInteger:
+    case (uint32_t)OSTypeKey::List:
+    case (uint32_t)OSTypeKey::Long:
+    case (uint32_t)OSTypeKey::RawData:
+    case (uint32_t)OSTypeKey::Reference:
+    case (uint32_t)OSTypeKey::UnitFloat:
+    case (uint32_t)OSTypeKey::String:
+        return true;
+    }
+    return false;
+}
+
+bool is_valid_reference_type(const uint32_t key) {
+    switch (key) {
+    case (uint32_t)OSTypeKey::RefClass:
+    case (uint32_t)OSTypeKey::RefEnum:
+    case (uint32_t)OSTypeKey::RefIdentifier:
+    case (uint32_t)OSTypeKey::RefIndex:
+    case (uint32_t)OSTypeKey::RefName:
+    case (uint32_t)OSTypeKey::RefOffset:
+    case (uint32_t)OSTypeKey::RefProperty:
+      return true;
+    }
+    return false;
+}
+
+bool is_valid_unit_float(const uint32_t unit) {
+  switch (unit) {
+  case (std::uint32_t)OSTypeUnitFloat::Unit::Angle:
+  case (std::uint32_t)OSTypeUnitFloat::Unit::Density:
+  case (std::uint32_t)OSTypeUnitFloat::Unit::Distance:
+  case (std::uint32_t)OSTypeUnitFloat::Unit::None:
+  case (std::uint32_t)OSTypeUnitFloat::Unit::Percent:
+  case (std::uint32_t)OSTypeUnitFloat::Unit::Pixel:
+    return true;
+  }
+  return false;
+}
+
+OSTypeClassMetaType Decoder::parseDescrVariable() {
+    auto const classIDLength = read32();
+    OSTypeClassMetaType meta;
+    if (classIDLength == 0) {
+        meta.keyClassID = read32();
+        TRACE("ClassID: %d", meta.keyClassID);
+    }
+    else {
+        meta.name.resize(classIDLength);
+        m_file->read((uint8_t*)&meta.name[0], classIDLength);
+        TRACE("ClassID: %s", meta.name.c_str());
+    }
+    return meta;
+}
+
+std::unique_ptr<OSType> Decoder::parseReferenceType() {
+    const uint32_t nItems = read32();
+    auto ref = std::make_unique<OSTypeReference>();
```

Comment:
To use `std::make_unique<>` we require C++14. I've switched to C++14 in `aseprite` repo [here](https://github.com/aseprite/aseprite/commit/8ae2a8bce55ac81160e6e3295c7d5bd6d17ece55) but something similar will be required in the psd CMakeLists.txt file.